### ISSUE TITLE
Hyperopt gives different results when using different number of jobs

### ISSUE THREAD

**Issue by Cristy94** [freqtrade/freqtrade#9524]
## Describe your environment

  * Operating system: Windows 11
  * Python Version: 3.10.7 (`python -V`)
  * CCXT version: 4.1.22 (`pip freeze | grep ccxt`)
  * Freqtrade Version: 2023.10 (`freqtrade -V` or `docker compose run --rm freqtrade -V` for Freqtrade running in docker)
  
Note: All issues other than enhancement requests will be closed without further comment if the above template is deleted or not filled out.

## Describe the problem:

Running hyperopt with multiple cores (20), not only doesn't run faster, but it gives different (worse) results.
It looks like the parallelization makes hyperopt miss many cases.

### Steps to reproduce:

  1. Run hyperopt with -j 1
  2. Run hyperopt with multiple jobs (e.g. 20)
  
### Observed Results:

  * What happened? -  Runtime was the same (around 3 mins) but results were different, where using multiple threads lead to a lot worse profit.
  * What did you expect to happen? Hyperopt runs faster and leads to same, determinstic results.

**Comment by stash86** [comment]
1. Hyperopt use bayesian statistic, which means it won't go through all possible combinations like brute search would be.
2. more jobs means more memory used. No one ever said that more jobs will always finish faster.
3. Read this https://botacademy.ddns.net/2023/09/23/myth-blowing-hyperopt-will-turn-any-strategies-to-rocket-strategies/ . What the bot sees as better result (due to the loss function used) might not satisfy your preference

**Comment by xmatthias** [comment]
Aside from the above, hyperopt is also using a different random state on each startup, to get you different combinations if you run the same hyperopt-run differently (this will allow hyperopt to find new minima's in it's bayesian search.

You can force a specific random state via `--random-state INT` - if you're after reproducible results.

> --random-state INT    Set random state to some positive integer for reproducible hyperopt results.

---

**edit WAY too long later** after thinking about this a bit more, this is expected to give different results IF different `-j` entries are used.
the difference is - with -j1, hyperopt will try to find the next best point in the multi-dimensional space after every backtest.

with `-j2` - it'll try to do the same thing only every other backtest - hence it will find different points, as it has 1 additional datapoint to analyze the next direction.

**Comment by Cristy94** [comment]
> 1. Hyperopt use bayesian statistic, which means it won't go through all possible combinations like brute search would be.
> 2. more jobs means more memory used. No one ever said that more jobs will always finish faster.
> 3. Read this https://botacademy.ddns.net/2023/09/23/myth-blowing-hyperopt-will-turn-any-strategies-to-rocket-strategies/ . What the bot sees as better result (due to the loss function used) might not satisfy your preference

1. Ok, so it also has a random starting seed, which might lead to different results when ran multiple times.
2. But if I have 24 cores, shouldn't it be a way to make hyperopt faster than running with 2 cores? There's no way to speed-up hyperopt by adding more threads?
3. Thanks for sharing, but that blog post is very basic (mostly explains what the hyperopt docs explain). The bot loss function was chosen to be ProfitDrawDownHyperOptLoss, and the result when using more threads was a lot worse (much worse profit and larger drawdown).

If multiple jobs don't increase performance (no multi-threading) and only run hyperopt multiple times, it should at least lead to better results, right? As using 10 jobs would try 10 times finding best params with different startup rand seed?

**Comment by froggleston** [comment]
1. Yes
2. Typically yes, but not in every use case.

The supposed optimum, i.e. "better results", is based on the path through the search space, which isn't guaranteed to be the same each time.

**Comment by xmatthias** [comment]
2) it'll depend. usually, you'll encounter more memory pressure with more threads (it's also not threads but processes, ...).
You'll however need to be rather careful as to WHEN the actual time is spent.
Hyperopt will 
* load the data once
* create some initial random points (30) - and then spin of as many parallel processes as given (usually "cpu count -1" - or whatever is given via `-j`).
* each process will then run a backtest. once finished, they'll need to wait until all in this "batch" are finished to calculate the next points based on current points (this is a limitation of the scikit-learn bayesian search algo, which is not parallelizable).
* calculate a new batch of points, re-peat the previous and this step until all epochs are done.

This last 2 steps is where the "parallel" time will be spent - but it's "parallel in batches" - with each process also waiting on the "slowest" backtest in each batch (you'll see this usually as spikes in the cpu utilization).

The _actual_ performance (gain) will depend quite a lot on the amount of parameters in your strategy - the more parameters you have, the more time will be spent in the "calculate new points" step - which is a single-core process (complain to scikit-learn if you feel like it ... ).

**Comment by Cristy94** [comment]
Thanks for the answer.

Is there a debugger mode? To show where time is being spent on? Like `freqtrade hyerpot --profiling=true`?


What is strange, is that the CPU cores are not 100% utilized, there seems to be a lot of waiting, I think not even one core is constantly at 100% usage for more than a few seconds.

> create some initial random points (30)

So all jobs start from the same initial seed/random points, or a random set for each job?

I think something feels off because the initial iteration (first N jobs) finish very quickly and show ok results, but then, the later the epoch, the more time an iteration takes. Aren't all iterations similar in terms of complexity?

**Comment by xmatthias** [comment]
there is only one job, hence a total of 30 points per `freqtrade hyperopt`.

There's no profiling feature - and i don't really see why there should be. 
Behavior of scikit-optimize is well documented (e.g. https://github.com/scikit-optimize/scikit-optimize/issues/768) - and has nothing really to do with how freqtrade operates, as we're just using that.

If you look at the CPU graph in that issue - you'll find huge similarities between your system usage and that. 
Sure, by having tons of cores, it'll look slightly different, most likely less "spiky" - but on a 4-8core system, that's more or less exactly what you'll get (obvoiusly depending on the strategy complexity / runtime for one backtest).

Also, the "intra-spike" times will increase with bigger spaces - as each additional parameter adds one dimension to the parameter space.

**Comment by Cristy94** [comment]
I took a look over the scikit issue, and I noticed scikit is no longer maintained/improved (last release was 2 years ago).

What about switching to a different hyperoptimization tool like https://github.com/bayesian-optimization/BayesianOptimization which as more stars and is also actively maintained?

I am pretty sure the hyperopt performance can be improved at least 10x with the right optimizer and multi-threading implementation/design.

**Comment by xmatthias** [comment]
Scikit-optimize is no longer maintained, but scikit-learn is (which is what's actually doing the heavy lifting).

Ad hock, i don't see how this would change anything over scikit-optimize - as it seems to be using the same, underlying scikit-learn algorithms.
I also don't find any performance comparison for this project (not even proper documentation beyond "these are some sample notebooks", and a wiki page that was last updated 2017...) - but i'm sure you've found some that justifies a statement of "least 10x improvement".

---

Feel free to invest the time / work and submit a pull request trying to switch out scikit-optimize (optuna might be an alternative version, which at a glance, seems a bit more mature / featurecomplete).

At any rate (independent of the framework) we do have some requirements for this 
* user facing interface should be kept unchanged as much as possible - for both strategy authors, as well as how you run hyperopt (that 2nd one is the easy part).
you probably can't ensure that for all advanced features hyperopt currently supports - but for what 90% of the users use, it should remain compatible.
* identical or improved performance (in both results in a multidimensional space) - as well as runtime


Looking for your pull request eventually ... :)

**Comment by Cristy94** [comment]
> Scikit-optimize is no longer maintained, but scikit-learn is (which is what's actually doing the heavy lifting).
> 
> Ad hock, i don't see how this would change anything over scikit-optimize - as it seems to be using the same, underlying scikit-learn algorithms. I also don't find any performance comparison for this project (not even proper documentation beyond "these are some sample notebooks", and a wiki page that was last updated 2017...) - but i'm sure you've found some that justifies a statement of "least 10x improvement".
> 
> Feel free to invest the time / work and submit a pull request trying to switch out scikit-optimize (optuna might be an alternative version, which at a glance, seems a bit more mature / featurecomplete).
> 
> At any rate (independent of the framework) we do have some requirements for this
> 
> * user facing interface should be kept unchanged as much as possible - for both strategy authors, as well as how you run hyperopt (that 2nd one is the easy part).
>   you probably can't ensure that for all advanced features hyperopt currently supports - but for what 90% of the users use, it should remain compatible.
> * identical or improved performance (in both results in a multidimensional space) - as well as runtime
> 
> Looking for your pull request eventually ... :)

The reason I said I'm pretty sure 10x improvement can be achieved is because the CPU cores are used only used around 10% of their capacity. So, simply making sure all of it is used, even without drastic algorithm changes, would result in this improvement.
Maybe it could be as simple as changing how the points are generated or tested, to use all cores. It could also maybe be something like having more independent optimizers, so they don't have to wait for batches to complete.

Sometimes this CPUs under-utilization occurs because data is not cached or transferred properly, so the CPU has to wait for RAM memory transfer, or even worse for disk transfer.

**Comment by xmatthias** [comment]
You clearly didn't read (or understand) the scikit-optimize issue i linked at. 
it's the fact that bayesian search isn't parallelizable that's causing the long "intermediate" times (assuming a decently fast strategy). Feel free to have a go at that - a PR to scikit-learn would be necessary for that to happen (but in my understanding, it's the algorithm itself that causes the non-parallelizability).

Running multiple parallel Optimizations won't help hyperopt.
while it will cause higher CPU utilization (you can also generate random numbers if that's what you're after) - that's the only thing it'll cause, while at the same time, producing identical (more likely worse) results.
If one optimization finds a great buy signal - and another finds a great sell signal - they'll never be combined, as the two won't know about each other - and the result may even be bad in both cases, as the "other" signal might be the worst possible scenario.


> Sometimes this CPUs under-utilization occurs because data is not cached or transferred properly, so the CPU has to wait for RAM memory transfer, or even worse for disk transfer.

Taking random guesses without understanding the problem won't help anyone.


---

At this point, i don't think there's much to add to this issue other than wild, random guesses (see last comment).

If you think you can improve performance - great, looking for your PR.
Please take good consideration into which alternative library you're picking (if any) - taking into account dependencies of this - as well as performance (speed and good results), ease of use, documentation and community size.

**Comment by xmatthias** [comment]
Small disclaimer - i've just had to edit my above comment about --random-state ( https://github.com/freqtrade/freqtrade/issues/9524#issuecomment-1849358456).

--random-state will NOT give the same results if `-j` is changed - as that will result in different batch-sizes, and therefore, different points in the hyperdimensional space (full explanation above under the "edit" section).

**Comment by Cristy94** [comment]
> Scikit-optimize is no longer maintained, but scikit-learn is (which is what's actually doing the heavy lifting).
> 
> Ad hock, i don't see how this would change anything over scikit-optimize - as it seems to be using the same, underlying scikit-learn algorithms. I also don't find any performance comparison for this project (not even proper documentation beyond "these are some sample notebooks", and a wiki page that was last updated 2017...) - but i'm sure you've found some that justifies a statement of "least 10x improvement".
> 
> Feel free to invest the time / work and submit a pull request trying to switch out scikit-optimize (optuna might be an alternative version, which at a glance, seems a bit more mature / featurecomplete).
> 
> At any rate (independent of the framework) we do have some requirements for this
> 
> * user facing interface should be kept unchanged as much as possible - for both strategy authors, as well as how you run hyperopt (that 2nd one is the easy part).
>   you probably can't ensure that for all advanced features hyperopt currently supports - but for what 90% of the users use, it should remain compatible.
> * identical or improved performance (in both results in a multidimensional space) - as well as runtime
> 
> Looking for your pull request eventually ... :)

Isn't skopt doing the heavy lifting? 99% of the time is spent in Optimizer.ask()
Would replacing/trying out something like https://github.com/optuna/optuna instead of skopt be hard?

**Comment by xmatthias** [comment]
> Would replacing/trying out something like https://github.com/optuna/optuna instead of skopt be hard?

I don't know optuna - so it's difficult to tell.
Learn it, give it a try ...

from past endavors in this direction (replacing a previous framework with scikit-optimize) i remember it was basically full-time work over about 2 weeks. 
it might be simpler than that - or it might be harder.
it might even fail as it's somehow incompatible with our requirements.
